scripts/lib/ts-generator.js

```diff
@@ -0,0 +1,437 @@
+TSGenerator.prototype.generateDocString = function generateDocString(documentation, tabCount) {
+    tabCount = tabCount || 0;
+    var code = '';
+    code += this.tabs(tabCount) + '/**\n';
+    code += this.tabs(tabCount) + ' * ' + this.transformDocumentation(documentation) + '\n';
+    code += this.tabs(tabCount) + ' */\n';
+    return code;
+};
+
+/**
+ * Returns an array of custom configuration options based on a service identiffier.
+ * Custom configuration options are determined by checking the metadata.json file.
+ */
+TSGenerator.prototype.generateCustomConfigFromMetadata = function generateCustomConfigFromMetadata(serviceIdentifier) {
+    // some services have additional configuration options that are defined in the metadata.json file
+    // i.e. dualstackAvailable = useDualstack
+    // create reference to custom options
+    var customConfigurations = [];
+    var serviceMetadata = this.metadata[serviceIdentifier];
+    // loop through metadata members
+    for (var memberName in serviceMetadata) {
+        if (!serviceMetadata.hasOwnProperty(memberName)) {
+            continue;
+        }
+        var memberValue = serviceMetadata[memberName];
+        // check configs
+        switch (memberName) {
+            case 'dualstackAvailable':
+                customConfigurations.push(CUSTOM_CONFIG_ENUMS.DUALSTACK);
+                break;
+        }
+    }
+
+    return customConfigurations;
+};
+
+/**
+ * Generates a type or interface based on the shape.
+ */
+TSGenerator.prototype.generateTypingsFromShape = function generateTypingsFromShape(shapeKey, shape, tabCount) {
+    // some shapes shouldn't be generated if they are javascript primitives
+    if (['string', 'boolean', 'number', 'Date', 'Blob'].indexOf(shapeKey) >= 0) {
+        return '';
+    }
+    var self = this;
+    var code = '';
+    tabCount = tabCount || 0;
+    var tabs = this.tabs;
+    var type = shape.type;
+    if (type === 'structure') {
+        code += tabs(tabCount) + 'export interface ' + shapeKey + ' {\n';
+        var members = shape.members;
+        // cycle through members
+        var memberKeys = Object.keys(members);
+        memberKeys.forEach(function(memberKey) {
+            // docs
+            var member = members[memberKey];
+            if (member.documentation) {
+                code += self.generateDocString(member.documentation, tabCount + 1);
+            }
+            var required = self.checkRequired(shape.required || [], memberKey) ? '?' : '';
+            code += tabs(tabCount + 1) + memberKey + required + ': ' + member.shape + ';\n';
+        });
+        code += tabs(tabCount) + '}\n';
+    } else if (type === 'list') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = ' + shape.member.shape + '[];\n';
+    } else if (type === 'map') {
+        code += tabs(tabCount) + 'export type ' + shapeKey + ' = {[key: string]: ' + shape.value.shape + '};\n';
+    } else if (type === 'string' || type === 'character') {
```

Comment:
It would be helpful to support enums.
(after checking for complex types and before starting to check primitive types, so between map and string):
```
else if (Array.isArray(shape.enum)) {
    code += tabs(tabCount) + 'export type ' + shapeKey + ' = ' + shape.enum.map(JSON.stringify).join('|') + ';\n';
```
So for example, Lambda's `Runtime` declaration would look like:
```
export type Runtime = "nodejs"|"nodejs4.3"|"java8"|"python2.7";
```
instead of just a generic string
```
export type Runtime = string;
```